// @ezvida/adl-core/flattener/SimpleArchetypeRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ArchetypeFlattener_1 = require("./ArchetypeFlattener");
var SimpleArchetypeRepository = /** @class */ (function () {
    function SimpleArchetypeRepository(archetypes) {
        this.archetypeIdDifferencialMap = new Map();
        try {
            for (var archetypes_1 = tslib_1.__values(archetypes), archetypes_1_1 = archetypes_1.next(); !archetypes_1_1.done; archetypes_1_1 = archetypes_1.next()) {
                var archetype = archetypes_1_1.value;
                this.archetypeIdDifferencialMap.set(archetype.archetypeId, archetype);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (archetypes_1_1 && !archetypes_1_1.done && (_a = archetypes_1.return)) _a.call(archetypes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.flattener = new ArchetypeFlattener_1.ArchetypeFlattener();
        var e_1, _a;
    }
    SimpleArchetypeRepository.prototype.parseAndFlattenArchetype = function (archetypeId) {
        var differential = this.getDifferentialArchetype(archetypeId);
        var parent;
        if (differential.parentArchetypeId != null) {
            parent = this.getFlatArchetype(differential.parentArchetypeId);
        }
        else {
            parent = null;
        }
        return this.flattener.flatten(parent, differential);
    };
    SimpleArchetypeRepository.prototype.getDifferentialArchetype = function (archetypeId) {
        return this.archetypeIdDifferencialMap.get(archetypeId);
    };
    SimpleArchetypeRepository.prototype.getFlatArchetype = function (archetypeId) {
        var flat = this.archetypeIdFlatMap.get(archetypeId);
        if (flat == null) {
            flat = this.parseAndFlattenArchetype(archetypeId);
            this.archetypeIdFlatMap.set(archetypeId, flat);
        }
        return flat;
    };
    return SimpleArchetypeRepository;
}());
exports.SimpleArchetypeRepository = SimpleArchetypeRepository;
