// @ezvida/adl-core/typed-json/customizer/encryption/RsaCryptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("constants");
var crypto = require("crypto");
var Element_1 = require("../../../rm/datastructures/Element");
var RsaCryptor = /** @class */ (function () {
    function RsaCryptor(privateKey, publicKey) {
        this.privateKey = privateKey;
        this.publicKey = publicKey;
    }
    RsaCryptor.prototype.getType = function () {
        return Element_1.Element;
    };
    RsaCryptor.prototype.onDeSerializer = function (json) {
        if (json.value) {
            var buffer = Buffer.from(json.value, 'base64');
            var dencrypted = crypto.privateDecrypt({ key: this.privateKey, padding: constants_1.RSA_PKCS1_PADDING }, buffer);
            json.value = JSON.parse(dencrypted.toString());
        }
        return json;
    };
    RsaCryptor.prototype.onSerializer = function (json) {
        if (json.value) {
            var buffer = Buffer.from(JSON.stringify(json.value), 'utf8');
            json.value = crypto.publicEncrypt({ key: this.publicKey, padding: constants_1.RSA_PKCS1_PADDING }, buffer);
        }
        return json;
    };
    return RsaCryptor;
}());
exports.RsaCryptor = RsaCryptor;
