// @ezvida/adl-core/rm/datavalues/quantity/ProportionKind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProportionKind;
(function (ProportionKind) {
    ProportionKind[ProportionKind["Ratio"] = 0] = "Ratio";
    ProportionKind[ProportionKind["Unitary"] = 1] = "Unitary";
    ProportionKind[ProportionKind["Percent"] = 2] = "Percent";
    ProportionKind[ProportionKind["Fraction"] = 3] = "Fraction";
    ProportionKind[ProportionKind["IntegerFraction"] = 4] = "IntegerFraction";
})(ProportionKind = exports.ProportionKind || (exports.ProportionKind = {}));
(function (ProportionKind) {
    function validProportionKind(nq) {
        switch (nq) {
            case ProportionKind.Ratio:
            case ProportionKind.Unitary:
            case ProportionKind.Percent:
            case ProportionKind.Fraction:
            case ProportionKind.IntegerFraction:
                return true;
            default:
                return false;
        }
    }
    ProportionKind.validProportionKind = validProportionKind;
})(ProportionKind = exports.ProportionKind || (exports.ProportionKind = {}));
