// @ezvida/adl-core/query/RmQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var jsonpath_1 = require("jsonpath");
var _ = require("lodash");
var RmQuery = /** @class */ (function () {
    function RmQuery(root) {
        this.POSITIONAL_PATTERN = /\/([a-z_\w]?(?:\w*))(?:\[(\d+)\])?/g;
        this.ROOT = '$';
        this.root = root;
    }
    RmQuery.prototype.findByAmPath = function (path) {
        return jsonpath_1.query(this.root, "$..[?(@.amPath=='" + path + "')]");
    };
    RmQuery.prototype.findByAPath = function (path) {
        return jsonpath_1.value(this.root, this.transform(path));
    };
    RmQuery.prototype.transform = function (path) {
        var syntax = this.ROOT;
        var match = this.POSITIONAL_PATTERN.exec(path);
        while (match != null) {
            var attribute = match[1];
            var index = match[2];
            var propertyKey = _.camelCase(attribute);
            syntax += "." + propertyKey + (index != null ? "[" + index + "]" : '');
            match = this.POSITIONAL_PATTERN.exec(path);
        }
        return syntax;
    };
    return RmQuery;
}());
exports.RmQuery = RmQuery;
