// antlr4ts/misc/Array2DHashMap.js
"use strict";
/*!
 * Copyright 2016 The ANTLR Project. All rights reserved.
 * Licensed under the BSD-3-Clause license. See LICENSE file in the project root for license information.
 */
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Array2DHashSet_1 = require("./Array2DHashSet");
var Stubs_1 = require("./Stubs");
var DefaultEqualityComparator_1 = require("./DefaultEqualityComparator");
var MapKeyEqualityComparator = /** @class */ (function () {
    function MapKeyEqualityComparator(keyComparator) {
        this.keyComparator = keyComparator;
    }
    MapKeyEqualityComparator.prototype.hashCode = function (obj) {
        return this.keyComparator.hashCode(obj.key);
    };
    MapKeyEqualityComparator.prototype.equals = function (a, b) {
        return this.keyComparator.equals(a.key, b.key);
    };
    return MapKeyEqualityComparator;
}());
var Array2DHashMap = /** @class */ (function () {
    function Array2DHashMap(keyComparer) {
        if (keyComparer instanceof Array2DHashMap) {
            this.backingStore = new Array2DHashSet_1.Array2DHashSet(keyComparer.backingStore);
        }
        else {
            this.backingStore = new Array2DHashSet_1.Array2DHashSet(new MapKeyEqualityComparator(keyComparer));
        }
    }
    Array2DHashMap.prototype.clear = function () {
        this.backingStore.clear();
    };
    Array2DHashMap.prototype.containsKey = function (key) {
        return this.backingStore.contains({ key: key });
    };
    Array2DHashMap.prototype.containsValue = function (value) {
        return this.values().contains(value);
    };
    Array2DHashMap.prototype.entrySet = function () {
        return new EntrySet(this, this.backingStore);
    };
    Array2DHashMap.prototype.get = function (key) {
        var bucket = this.backingStore.get({ key: key });
        if (!bucket) {
            return undefined;
        }
        return bucket.value;
    };
    Object.defineProperty(Array2DHashMap.prototype, "isEmpty", {
        get: function () {
            return this.backingStore.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    Array2DHashMap.prototype.keySet = function () {
        return new KeySet(this, this.backingStore);
    };
    Array2DHashMap.prototype.put = function (key, value) {
        var element = this.backingStore.get({ key: key, value: value });
        var result;
        if (!element) {
            this.backingStore.add({ key: key, value: value });
        }
        else {
            result = element.value;
            element.value = value;
        }
        return result;
    };
    Array2DHashMap.prototype.putIfAbsent = function (key, value) {
        var element = this.backingStore.get({ key: key, value: value });
        var result;
        if (!element) {
            this.backingStore.add({ key: key, value: value });
        }
        else {
            result = element.value;
        }
        return result;
    };
    Array2DHashMap.prototype.putAll = function (m) {
        try {
            for (var _a = __values(Stubs_1.asIterable(m.entrySet())), _b = _a.next(); !_b.done; _b = _a.next()) {
                var entry = _b.value;
                this.put(entry.getKey(), entry.getValue());
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var e_1, _c;
    };
    Array2DHashMap.prototype.remove = function (key) {
        var value = this.get(key);
        this.backingStore.remove({ key: key });
        return value;
    };
    Object.defineProperty(Array2DHashMap.prototype, "size", {
        get: function () {
            return this.backingStore.size;
        },
        enumerable: true,
        configurable: true
    });
    Array2DHashMap.prototype.values = function () {
        return new ValueCollection(this, this.backingStore);
    };
    Array2DHashMap.prototype.hashCode = function () {
        return this.backingStore.hashCode();
    };
    Array2DHashMap.prototype.equals = function (o) {
        if (!(o instanceof Array2DHashMap)) {
            return false;
        }
        return this.backingStore.equals(o.backingStore);
    };
    return Array2DHashMap;
}());
exports.Array2DHashMap = Array2DHashMap;
var EntrySet = /** @class */ (function () {
    function EntrySet(map, backingStore) {
        this.map = map;
        this.backingStore = backingStore;
    }
    EntrySet.prototype.add = function (e) {
        throw new Error("Not implemented");
    };
    EntrySet.prototype.addAll = function (collection) {
        throw new Error("Not implemented");
    };
    EntrySet.prototype.clear = function () {
        this.map.clear();
    };
    EntrySet.prototype.contains = function (o) {
        throw new Error("Not implemented");
    };
    EntrySet.prototype.containsAll = function (collection) {
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                if (!this.contains(key)) {
                    return false;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return true;
        var e_2, _c;
    };
    EntrySet.prototype.equals = function (o) {
        if (o === this) {
            return true;
        }
        else if (!(o instanceof EntrySet)) {
            return false;
        }
        return this.backingStore.equals(o.backingStore);
    };
    EntrySet.prototype.hashCode = function () {
        return this.backingStore.hashCode();
    };
    Object.defineProperty(EntrySet.prototype, "isEmpty", {
        get: function () {
            return this.backingStore.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    EntrySet.prototype.iterator = function () {
        throw new Error("Not implemented");
    };
    EntrySet.prototype.remove = function (o) {
        throw new Error("Not implemented");
    };
    EntrySet.prototype.removeAll = function (collection) {
        var removedAny = false;
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                removedAny = this.remove(key) || removedAny;
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return removedAny;
        var e_3, _c;
    };
    EntrySet.prototype.retainAll = function (collection) {
        throw new Error("Not implemented");
    };
    Object.defineProperty(EntrySet.prototype, "size", {
        get: function () {
            return this.backingStore.size;
        },
        enumerable: true,
        configurable: true
    });
    EntrySet.prototype.toArray = function (a) {
        throw new Error("Not implemented");
    };
    return EntrySet;
}());
var KeySet = /** @class */ (function () {
    function KeySet(map, backingStore) {
        this.map = map;
        this.backingStore = backingStore;
    }
    KeySet.prototype.add = function (e) {
        throw new Error("Not supported");
    };
    KeySet.prototype.addAll = function (collection) {
        throw new Error("Not supported");
    };
    KeySet.prototype.clear = function () {
        this.map.clear();
    };
    KeySet.prototype.contains = function (o) {
        return this.backingStore.contains({ key: o });
    };
    KeySet.prototype.containsAll = function (collection) {
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                if (!this.contains(key)) {
                    return false;
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return true;
        var e_4, _c;
    };
    KeySet.prototype.equals = function (o) {
        if (o === this) {
            return true;
        }
        else if (!(o instanceof KeySet)) {
            return false;
        }
        return this.backingStore.equals(o.backingStore);
    };
    KeySet.prototype.hashCode = function () {
        return this.backingStore.hashCode();
    };
    Object.defineProperty(KeySet.prototype, "isEmpty", {
        get: function () {
            return this.backingStore.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    KeySet.prototype.iterator = function () {
        throw new Error("Not implemented");
    };
    KeySet.prototype.remove = function (o) {
        return this.backingStore.remove({ key: o });
    };
    KeySet.prototype.removeAll = function (collection) {
        var removedAny = false;
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                removedAny = this.remove(key) || removedAny;
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_5) throw e_5.error; }
        }
        return removedAny;
        var e_5, _c;
    };
    KeySet.prototype.retainAll = function (collection) {
        throw new Error("Not implemented");
    };
    Object.defineProperty(KeySet.prototype, "size", {
        get: function () {
            return this.backingStore.size;
        },
        enumerable: true,
        configurable: true
    });
    KeySet.prototype.toArray = function (a) {
        throw new Error("Not implemented");
    };
    return KeySet;
}());
var ValueCollection = /** @class */ (function () {
    function ValueCollection(map, backingStore) {
        this.map = map;
        this.backingStore = backingStore;
    }
    ValueCollection.prototype.add = function (e) {
        throw new Error("Not supported");
    };
    ValueCollection.prototype.addAll = function (collection) {
        throw new Error("Not supported");
    };
    ValueCollection.prototype.clear = function () {
        this.map.clear();
    };
    ValueCollection.prototype.contains = function (o) {
        try {
            for (var _a = __values(Stubs_1.asIterable(this.backingStore)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var bucket = _b.value;
                if (DefaultEqualityComparator_1.DefaultEqualityComparator.INSTANCE.equals(o, bucket.value)) {
                    return true;
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_6) throw e_6.error; }
        }
        return false;
        var e_6, _c;
    };
    ValueCollection.prototype.containsAll = function (collection) {
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                if (!this.contains(key)) {
                    return false;
                }
            }
        }
        catch (e_7_1) { e_7 = { error: e_7_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_7) throw e_7.error; }
        }
        return true;
        var e_7, _c;
    };
    ValueCollection.prototype.equals = function (o) {
        if (o === this) {
            return true;
        }
        else if (!(o instanceof ValueCollection)) {
            return false;
        }
        return this.backingStore.equals(o.backingStore);
    };
    ValueCollection.prototype.hashCode = function () {
        return this.backingStore.hashCode();
    };
    Object.defineProperty(ValueCollection.prototype, "isEmpty", {
        get: function () {
            return this.backingStore.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    ValueCollection.prototype.iterator = function () {
        var delegate = this.backingStore.iterator();
        return {
            hasNext: function () {
                return delegate.hasNext();
            },
            next: function () {
                return delegate.next().value;
            },
            remove: function () {
                throw new Error("Not supported");
            }
        };
    };
    ValueCollection.prototype.remove = function (o) {
        throw new Error("Not implemented");
    };
    ValueCollection.prototype.removeAll = function (collection) {
        var removedAny = false;
        try {
            for (var _a = __values(Stubs_1.asIterable(collection)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                removedAny = this.remove(key) || removedAny;
            }
        }
        catch (e_8_1) { e_8 = { error: e_8_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_8) throw e_8.error; }
        }
        return removedAny;
        var e_8, _c;
    };
    ValueCollection.prototype.retainAll = function (collection) {
        throw new Error("Not implemented");
    };
    Object.defineProperty(ValueCollection.prototype, "size", {
        get: function () {
            return this.backingStore.size;
        },
        enumerable: true,
        configurable: true
    });
    ValueCollection.prototype.toArray = function (a) {
        if (a === undefined || a.length < this.backingStore.size) {
            a = new Array(this.backingStore.size);
        }
        var i = 0;
        try {
            for (var _a = __values(Stubs_1.asIterable(this.backingStore)), _b = _a.next(); !_b.done; _b = _a.next()) {
                var bucket = _b.value;
                a[i++] = bucket.value;
            }
        }
        catch (e_9_1) { e_9 = { error: e_9_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_9) throw e_9.error; }
        }
        return a;
        var e_9, _c;
    };
    return ValueCollection;
}());
